tools/doc-generator/main.go

```diff
@@ -0,0 +1,239 @@
+package main
+
+import (
+	"fmt"
+	"os"
+	"reflect"
+	"strings"
+
+	"github.com/cortexproject/cortex/pkg/alertmanager"
+	"github.com/cortexproject/cortex/pkg/chunk"
+	"github.com/cortexproject/cortex/pkg/chunk/cache"
+	"github.com/cortexproject/cortex/pkg/chunk/storage"
+	config_client "github.com/cortexproject/cortex/pkg/configs/client"
+	"github.com/cortexproject/cortex/pkg/configs/db"
+	"github.com/cortexproject/cortex/pkg/cortex"
+	"github.com/cortexproject/cortex/pkg/distributor"
+	"github.com/cortexproject/cortex/pkg/ingester"
+	"github.com/cortexproject/cortex/pkg/ingester/client"
+	"github.com/cortexproject/cortex/pkg/querier"
+	"github.com/cortexproject/cortex/pkg/querier/frontend"
+	"github.com/cortexproject/cortex/pkg/querier/queryrange"
+	"github.com/cortexproject/cortex/pkg/ring/kv/consul"
+	"github.com/cortexproject/cortex/pkg/ring/kv/etcd"
+	"github.com/cortexproject/cortex/pkg/ring/kv/memberlist"
+	"github.com/cortexproject/cortex/pkg/ruler"
+	"github.com/cortexproject/cortex/pkg/util/validation"
+	"github.com/weaveworks/common/server"
+)
+
+const (
+	maxLineWidth = 80
+	tabWidth     = 2
+)
+
+var (
+	// Ordered list of root blocks. The order is the same order that will
+	// follow the markdown generation.
+	rootBlocks = []rootBlock{
```

Comment:
It would be nicer if this information was directly in the Cortex code, if possible. One idea is to put it into cortex.go, next to the reference to config structure.

```
Server         server.Config            `yaml:"server,omitempty"` `doc:"name:server_config,configures the HTTP and gRPC server of the launched service(s)."`
```

(if tag syntax allows this at all)

Not sure how difficult it would be. But it would be very cool if we could reuse this for Loki as well, without too many modifications.

Another reason for doing it is that currently it's difficult to find these extra documentation lines. Tool itself is the last place where I would look for it.